src/cobo/handshake.c

```diff
@@ -998,25 +1073,29 @@ static int key_decrypt_packet(unsigned char *key, unsigned int key_len,
 #endif
 }
 
-static int compare_packets(handshake_packet_t *expected_packet,
-                           handshake_packet_t *recvd_packet)
+static int compare_packets(handshake_packet_t *expected_packet, handshake_packet_t *recvd_packet,
```

Comment:
Let's break this into two functions: compare_handshake_packets() and compare_random_number_packets().
Then compare_handshake_packets() after the first exchange, but before doing the random packet exchange. 
Then compare_random_number_packets() after the random packet exchange.

Two reasons for this. A) It's a common occurrence for two different spindle sessions to connect to each other and realize they are the wrong sessions (the HSHAKE_DROP_CONNECTION test below). Let's get those dropped quickly before sending the other packets. B) No reason to be signing and sending those other packets when we can do an earlier test for common cases..